### PR TITLE
introduce cpe as a type, added validation for cpe check

### PR DESCRIPTION
**Pull request by kchetans**
Define CPE as a type for SBOM, while reading CPE from CDX and SPDX we are validating if CPE format is valid or not.

### CODE REVIEW

File: pkg/sbom/spdx.go [hunk 1]
```diff
@@ -328,11 +329,21 @@ func (s *spdxDoc) cpes(index int) []string {
...
+			if cpeValidae.Valid() {
+				urls = append(urls, cpeValidae)
+			} else {
+				s.addToLogs(fmt.Sprintf("spdx doc pkg %s at index %d invalid cpes found", pkg.PackageName, index))
+			}
 		}
 
 		if strings.ToLower(p.RefType) == spdx_common.TypeSecurityCPE22Type {
-			urls = append(urls, p.Locator)
+			cpeValidae := cpe.NewCPE(p.Locator)
```
Comment (riteshnoronha) [inline pkg/sbom/spdx.go:341]: I believe this package only checks for valid 2.3 cpe not 2.2 can u clarify ?

Comment (riteshnoronha) [reply pkg/sbom/spdx.go:341]: lets rename cpeValidae to cpeV or something shorter.

File: pkg/sbom/spdx.go [hunk 2]
```diff
@@ -328,11 +329,21 @@ func (s *spdxDoc) cpes(index int) []string {
 
 	for _, p := range pkg.PackageExternalReferences {
 		if strings.ToLower(p.RefType) == spdx_common.TypeSecurityCPE23Type {
```
Comment (riteshnoronha) [inline pkg/sbom/spdx.go:331]: We need to add test cases for these functions.

File: pkg/sbom/cdx.go [hunk 3]
```diff
@@ -186,7 +187,12 @@ func (c *cdxDoc) parseComps() {
 		nc.purpose = string(sc.Type)
 		nc.isReqFieldsPresent = c.pkgRequiredFields(index)
 		nc.purls = []string{sc.PackageURL}
-		nc.cpes = []string{sc.CPE}
+		ncpe := cpe.NewCPE(sc.CPE)
```
Comment (riteshnoronha) [inline pkg/sbom/cdx.go:190]: Need to add test cases

File: pkg/cpe/cpe.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+type CPEMeta interface {
+	Valid() bool
+	String() string
+}
+
+// This regex string is taken from
+// https://csrc.nist.gov/schema/cpe/2.3/cpe-naming_2.3.xsd which has the official cpe spec
+// This first part matches CPE urls and the second part matches binding strings
+const cpeRegex = ((`^([c][pP][eE]:/[AHOaho]?(:[A-Za-z0-9\._\-~%]*){0,6})`) +
```
Comment (riteshnoronha) [inline pkg/cpe/cpe.go:30]: Need 2.3 & 2.2 support

**Review by riteshnoronha — CHANGES_REQUESTED** [review]

File: pkg/cpe/cpe.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+// limitations under the License.
+package cpe
+
+import (
+	"regexp"
+)
+
+type CPE string
+
+type CPEMeta interface {
```
Comment (riteshnoronha) [inline pkg/cpe/cpe.go:22]: There is no reason for interface.

**Review by riteshnoronha — APPROVED** [review]
Looks good to me, please remove the CPEMETA interface its not required.

**Review by riteshnoronha — DISMISSED** [review]

**Review by surendrapathak — APPROVED** [review]
Squashed commit 
LGTM

**Pull request merged**